domain/removal/service/unit.go

```diff
@@ -150,12 +156,25 @@ func (s *Service) processUnitRemovalJob(ctx context.Context, job removal.Job) er
 		return errors.Errorf("unit %q is alive", job.EntityUUID).Add(removalerrors.EntityStillAlive)
 	}
 
+	// If the unit is the leader of an application, we need to revoke
+	// leadership before we can delete it.
+	applicationName, unitName, err := s.st.GetApplicationNameAndUnitNameByUnitUUID(ctx, job.EntityUUID)
+	if err != nil {
+		return errors.Errorf("getting application name and unit name: %w", err)
+	}
+
+	if err := s.leadershipRevoker.RevokeLeadership(applicationName, unit.Name(unitName)); err != nil && !errors.Is(err, leadership.ErrClaimNotHeld) {
+		return errors.Errorf("revoking leadership: %w", err)
+	}
```

Comment:
How do we guarantee that another unit has successfully claimed the lease before deleting it?

Reply:
Shouldn't this be somehow in the same transaction as the unit delete?